server/rollback_runner.go

```diff
@@ -0,0 +1,156 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/bytebase/bytebase/api"
+	"github.com/bytebase/bytebase/common/log"
+	"github.com/bytebase/bytebase/plugin/db"
+	"github.com/bytebase/bytebase/plugin/db/mysql"
+)
+
+var (
+	generateRollbackSQLChan = make(chan *api.Task, 100)
+)
+
+// NewRollbackRunner creates a new rollback runner.
+func NewRollbackRunner(server *Server) *RollbackRunner {
+	return &RollbackRunner{
+		server: server,
+	}
+}
+
+// RollbackRunner is the rollback runner generating rollback SQL statements.
+type RollbackRunner struct {
+	server *Server
+}
+
+// Run starts the rollback runner.
+func (r *RollbackRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+	r.retryGetRollbackSQL(ctx)
+	for {
+		select {
+		case task := <-generateRollbackSQLChan:
+			r.getRollbackSQL(ctx, task)
+		case <-ctx.Done(): // if cancel() execute
+			return
+		}
+	}
+}
+
+// retryGetRollbackSQL retries generating rollback SQL for tasks.
+// It is currently called when Bytebase server starts and only rerun unfinished generation.
+func (r *RollbackRunner) retryGetRollbackSQL(ctx context.Context) {
+	find := &api.TaskFind{
+		StatusList: &[]api.TaskStatus{api.TaskRunning},
+		TypeList:   &[]api.TaskType{api.TaskDatabaseDataUpdate},
+		Payload:    "payload->>'threadID'!='' AND payload->>'rollbackError'='' AND payload->>'rollbackStatement'=''",
+	}
+	taskList, err := r.server.store.FindTask(ctx, find, true)
+	if err != nil {
+		log.Error("Failed to get running DML tasks", zap.Error(err))
+		return
+	}
+	for _, task := range taskList {
+		if task.Instance.Engine != db.MySQL {
```

Comment:
Is this still needed? Or make a comment saying this is a double checking.

Reply:
Not really, removed.